### PR TITLE
[api] Remove Misctypes

### PR DESCRIPTION
**Pull request by ejgallego**
The `Misctypes` file contains a popurri of type definitions, however, this diverse nature makes it impossible to map it to a concrete component, and kind of forces an obscure nature on it.

We relocate the definitions in Misctypes to other modules; for the most part, I think the relocation makes sense, removing this catch-all file from the tree.

### CODE REVIEW

**Review by gares — APPROVED** [review]
Misc = Mess

**Comment by gares** [comment]
I think your approach with a compat module full of deprecations is the way to go

**Comment by ejgallego** [comment]
I think that most of the types in `Misctypes` pertain to @ppedrot 's domain [tho a few additional reviews would be great], so I am proposing him to be the assignee of this PR.

**Review by maximedenes — APPROVED** [review]
I reviewed only the kernel changes, they look ok to me.

File: kernel/names.mli [hunk 1]
```diff
@@ -605,3 +605,9 @@ type evaluable_global_reference =
   | EvalConstRef of Constant.t
 
 val eq_egr : evaluable_global_reference ->  evaluable_global_reference -> bool
+
+(** Located identifiers and objects with syntax. *)
+
+type lident = Id.t CAst.t
```
Comment (ppedrot) [inline kernel/names.mli]: I don't think that these types should be put in the kernel. They refer to AST types that live well upwards.

File: engine/evarutil.mli [hunk 2]
```diff
@@ -22,12 +22,22 @@ open EConstr
 val new_meta : unit -> metavariable
 val mk_new_meta : unit -> constr
 
+(** General evar naming using intro patterns  *)
+type intro_pattern_naming_expr =
+  | IntroIdentifier of Id.t
+  | IntroFresh of Id.t
+  | IntroAnonymous
+
```
Comment (ppedrot) [inline engine/evarutil.mli]: This should probably be put into Namegen, but I don't have a strong opinion on that.

**Review by ppedrot — CHANGES_REQUESTED** [review]
Everything is mostly OK, except for the fact that the located versions of AST datatypes were put into the kernel. I really don't think it is a good idea, and this seems reflected by the fact they're not used that down.

File: kernel/names.mli (hunk 1 above)
Comment (ejgallego) [reply kernel/names.mli]: Where do you want to put them? Names is not such a bad place after all I think for them.

File: engine/evarutil.mli (hunk 2 above)
Comment (ejgallego) [reply engine/evarutil.mli]: Sounds reasonable, let me see how things would look

File: kernel/names.mli (hunk 1 above)
Comment (ejgallego) [reply kernel/names.mli]: We are talking like 4 lines; in fact I'd even merge `Names` and `Nameops` and get rid of one module. We have too many modules.

Comment (ejgallego) [reply kernel/names.mli]: Imagine having all identifier-related stuff in a single module, can't be true!

Comment (ppedrot) [reply kernel/names.mli]: Names in already too big for its own good, IMHO... I'd rather split it.

Comment (ejgallego) [reply kernel/names.mli]: I hope we are not leading to a world where 20% of the TCB are `open Foo` statements XD

Comment (ejgallego) [reply kernel/names.mli]: I'd rather have these 4 silly lines not block this PR for very long. I still think that `Names` is not such a bad place for the located objects as IMHO it seems natural that the module defining `Id.t` also defines it's located version.

Noted that this is already used in the next layer `library/` so this is far from "well upwards" but instead it is "immediatedly upwards".

Comment (ppedrot) [reply kernel/names.mli]: OK, OK, fine with me.

**Review by ppedrot — APPROVED** [review]

File: kernel/names.mli (hunk 1 above)
Comment (ejgallego) [reply kernel/names.mli]: I understand your concern, but couldn't do better :( but yeah, if we find another place I'll be very happy to move it. Note indeed that Names and Subst have an special place within the kernel as they are used by both the checker and the kernel itself. So in fact they should be placed in a different place, even in clib. We have loads of stuff there that the kernel uses so indeed that's a tricky question.

**Comment by ppedrot** [comment]
I'll leave a few days for other reviewers to comment on this PR. If they stay silent though, this will be considered as tacit acceptance.

File: engine/evarutil.mli (hunk 2 above)
Comment (ejgallego) [reply engine/evarutil.mli]: I looks better indeed, pushing the change to `Namegen`.

**Comment by ejgallego** [comment]
So I another small pass; IMHO it is not perfect but I think it is a good step. I do agree with the `lid` etc... comment, however we are talking about 3 lines and indeed the kernel could have some use for these structures so let's keep it there for now.

**Pull request merged**